File: go/problems/122_best_time_to_buy_and_sell_stock_ii.go
package problems

func MaxProfit_122(prices []int) int {
	var res int
	for i := 1; i < len(prices); i++ {
		res += max(0, prices[i]-prices[i-1])
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: rs/src/bin/122_best_time_to_buy_and_sell_stock_ii.rs
fn main() {
    let v = vec![7, 1, 5, 3, 6, 4];
    println!("{:?}", max_profit(v));

    let v = vec![1, 2, 3, 4, 5];
    println!("{:?}", max_profit(v));

    let v = vec![7, 6, 4, 3, 1];
    println!("{:?}", max_profit(v));

    let v = vec![2, 1, 2, 0, 1];
    println!("{:?}", max_profit(v));
}

fn max_profit(prices: Vec<i32>) -> i32 {
    let n = prices.len();
    let (mut dp0, mut dp1) = (0, -prices[0]);

    for i in prices.iter().take(n).skip(1) {
        dp0 = dp0.max(dp1 + i);
        dp1 = dp1.max(dp0 - i);
    }

    dp0
}
